packages/components/src/custom-select-control/index.js

```diff
@@ -257,3 +258,16 @@ export default function CustomSelectControl( props ) {
 		</div>
 	);
 }
+
+function CustomSelectControl( props ) {
+	return (
+		<ExperimentalCustomSelectControl
+			{ ...props }
+			__experimentalShowSelectedHint={ false }
+		/>
+	);
+}
+
+lock( CustomSelectControl, ExperimentalCustomSelectControl );
```

Comment:
Should we "lock" the experimental component within the stable one or have a unique "experiments" object that holds all the experimental components/functions of a given package? 

While in this case, these two are related, sometimes we might not having anything "obvious" to attach too (like a new experimental component...) so I wonder if we should be consistent and have some guideline to attach all top level items to a single "experiments" object?

Reply:
To visualise, it's between these two:

```js
import { experiments as blockEditorExperiments, CustomSelectControl } from '@wordpress/block-editor';
import { unlock } from './experiments';

const ExperimentalCustomSelectControl = unlock( CustomSelectControl );
const { topLevelFunction } = unlock( blockEditorExperiments );
```

vs 

```js
import { experiments as blockEditorExperiments } from '@wordpress/block-editor';
import { unlock } from './experiments';

const { ExperimentalCustomSelectControl } = unlock( experiments );
const { topLevelFunction } = unlock( blockEditorExperiments );
```

I like the first one better:

* It feels consistent with how a store can be unlocked (`unlock( select( store ) )`)
* It reflects the strong coupling between the private and the public component
* It forces loading the specific component file which is handy for async loading

Also cc @jsnajdr who came up with the idea of locking.

Reply:
I like the second one personally because it's needed anyway for new components, APIs that are not tied to anything but the package and it's consistent. With the first one, we're going to be asking ourselves the same question and creating inconsistencies as we add locked APIs

Reply:
I'd normally consider the second one more heavily given the manual work involved in creating and maintaining all these experimental points of contact, but given that part of the goal here is to make this stuff as annoying as possible, I think the first isn't so bad.

is the problem that we're running this code on module initialization? that it's in the module global scope vs. in some kind of `init()` function? if so, could we _lock the lock_ behind a static latch?

```js
export let moduleAlreadyUnlocked = false;
```

sorry for not being very coherent here; I may not be that helpful on this issue.

Reply:
Something else I wondered re. conventions is if we should still use the `__experimental` prefix for a prop or a component that is locked?

Reply:
>because it's needed anyway for new components, APIs that are not tied to anything but the package and it's consistent.

That sounds right. My understanding([based on docs](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/coding-guidelines.md#experimental-functions-classes-and-variables)) is that we should already do something like that by creating an object to attach the new ones and then export to unlock. Unless I'm wrong :) 

>Something else I wondered re. conventions is if we should still use the __experimental prefix for a prop or a component that is locked?

If we have them locked I don't think it's that important. In some cases we might not be able to avoid it(components), but I guess we can lose them for props, etc.. 🤔

Reply:
> the goal here is to make this stuff as annoying as possible, I think the first isn't so bad.

The goal is to make it annoying for consumers, not for producers of these APIs. This impacts mostly producers and how we organize our code. Not sure it has any impact on whether it's annoying to consume it.

> prefix

Agree here that there's no need for prefixes anymore.

Reply:
The consensus seems to be with locking a top-level module export so move forward with that 👍

Reply:
I locked the component onto a top-level `experiments` object and and removed the `__experimental` prefix.

Reply:
Can't speak for the other packages, but at least for the wp-components case, I think there is still a strong need for a `__experimental` prefix because otherwise it's hard to tell visually/programatically whether something is a locked prop.

Some kind of note in a JSDoc would be better than nothing, but I think we will run into a lot of confusion if the prop name isn't clearly prefixed. In fact, I'd even prefer a *different* prefix for locked experimental props vs. non-locked (grandfathered) experimental props, because as a consumer I need to know whether I'll have to do an explicit `unlock` to use it. And by "consumer" I am mostly talking about devs in this repo, so the "it's experimental so we don't have to document it, nobody should be using it in the first place" argument doesn't apply.

When somebody looks at the Storybook props table, it must be very clear that a particular prop is experimental and/or requires explicit unlocking.

Reply:
Fair enough. OK, putting back the `__experimental` prefix.